src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -616,6 +628,7 @@ Pipe MergeTask::VerticalMergeStage::createPipeForReadingOneColumn(const String &
             *global_ctx->data,
             global_ctx->storage_snapshot,
             global_ctx->future_part->parts[part_num],
+            global_ctx->alter_conversions[part_num],
```

Comment:
I didn't get how it will avoid indexing because we need to store a vector of alter conversions anyways.  But it will required injecting context and metadata snapshot to `FutureMergedMutatedPart` which may be inconvenient in some places.